File: database-driver-plugin/src/main/java/de/lino/database/database/nosql/mongodb/MongoDBDatabaseProvider.java
package de.lino.database.database.nosql.mongodb;

import com.mongodb.client.MongoClient;
import com.mongodb.client.MongoClients;
import com.mongodb.client.MongoDatabase;
import de.lino.database.database.AbstractCachedDatabaseSection;
import de.lino.database.database.AbstractLazyDatabaseProvider;
import de.lino.database.database.SectionConfig;
import de.lino.database.database.auth.Credentials;
import de.lino.database.database.DatabaseProvider;
import de.lino.database.database.DatabaseSection;
import org.jetbrains.annotations.NotNull;

import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.text.MessageFormat;
import java.util.List;
import java.util.function.Consumer;

/**
 * The {@link DatabaseProvider} backed by a MongoDB database, each {@link DatabaseSection} a
 * collection via {@link MongoDBDatabaseSection}. Section lifecycle and caching live in
 * {@link AbstractLazyDatabaseProvider}; this class only supplies the collection-level storage
 * operations - listing collections, constructing a {@link MongoDBDatabaseSection}, dropping a
 * collection. {@link MongoClient} and {@link MongoDatabase} are themselves thread-safe and
 * designed for concurrent multi-threaded use, so every method here is safe to call
 * concurrently without additional locking.
 */
public class MongoDBDatabaseProvider extends AbstractLazyDatabaseProvider {

    /**
     * Collection names that are never exposed as a {@link DatabaseSection}, since they are
     * MongoDB-internal rather than application data.
     */
    private static final List<String> FORBIDDEN = List.of("system.version", "system.users");

    /**
     * The client connection this database and every section it creates share.
     */
    private final MongoClient mongoClient;

    /**
     * The database this database is connected to.
     */
    private final MongoDatabase mongoDatabase;

    /**
     * Connects to a MongoDB database with {@code credentials} and discovers every existing,
     * non-{@link #FORBIDDEN} collection's name. Only names - no section objects, no documents -
     * so construction cost is one collection listing, independent of how much the database
     * holds.
     *
     * @param credentials the login credentials and connection details to connect with
     */
    public MongoDBDatabaseProvider(@NotNull Credentials credentials) {

        this.mongoClient = MongoClients.create(MessageFormat.format(
                "mongodb://{0}:{1}@{2}:{3}/{4}",
                credentials.getUserName(),
                URLEncoder.encode(credentials.getPassword(), StandardCharsets.UTF_8),
                credentials.getAddress(),
                Integer.toString(credentials.getPort()),
                credentials.getDatabase()
        ));

        this.mongoDatabase = this.mongoClient.getDatabase(credentials.getDatabase());

        this.reload();

    }

    @Override
    public void shutdown() {
        this.mongoClient.close();
        this.forgetSections();
    }

    /**
     * {@inheritDoc}
     * <p>
     * Lists the database's collection names, skipping the {@link #FORBIDDEN} MongoDB-internal
     * ones.
     */
    @Override
    protected void discoverNames(@NotNull Consumer<String> consumer) {

        for (final String name : this.mongoDatabase.listCollectionNames()) {
            if (FORBIDDEN.contains(name)) continue;
            consumer.accept(name);
        }

    }

    @Override
    protected AbstractCachedDatabaseSection constructSection(@NotNull String name, @NotNull SectionConfig config) {
        return new MongoDBDatabaseSection(this.mongoDatabase, name, config);
    }

    @Override
    protected void dropSectionRemote(@NotNull String name) {
        this.mongoDatabase.getCollection(name).drop();
    }

}

File: python/database-driver-plugin/src/database_driver/plugin/database/nosql/mongodb/mongodb_database_provider.py
"""Mirror of ``de.lino.database.database.nosql.mongodb.MongoDBDatabaseProvider``."""

from __future__ import annotations

import urllib.parse
from collections.abc import Callable

from database_driver.api.database.auth.credentials import Credentials
from database_driver.api.database.section_config import SectionConfig
from database_driver.plugin.database.abstract_cached_database_section import AbstractCachedDatabaseSection
from database_driver.plugin.database.abstract_lazy_database_provider import AbstractLazyDatabaseProvider
from database_driver.plugin.database.nosql.mongodb.mongodb_database_section import MongoDBDatabaseSection

_FORBIDDEN = ["system.version", "system.users"]
"""Collection names that are never exposed as a section, since they are MongoDB-internal
rather than application data."""


class MongoDBDatabaseProvider(AbstractLazyDatabaseProvider):
    """The ``DatabaseProvider`` backed by a MongoDB database, each section a collection
    via ``MongoDBDatabaseSection``. Section lifecycle and caching live in
    ``AbstractLazyDatabaseProvider``; this class only supplies the collection-level
    storage operations - listing collections, constructing a section, dropping a
    collection. pymongo's ``MongoClient`` is itself thread-safe and designed for
    concurrent multi-threaded use, so every method here is safe to call concurrently
    without additional locking.

    Connects with ``credentials`` and discovers every existing, non-internal
    collection's name. Only names - no section objects, no documents - so construction
    cost is one collection listing, independent of how much the database holds.
    """

    def __init__(self, credentials: Credentials) -> None:
        super().__init__()

        from pymongo import MongoClient

        self.mongo_client = MongoClient(
            "mongodb://{user}:{password}@{address}:{port}/{database}".format(
                user=credentials.user_name,
                password=urllib.parse.quote(credentials.password, safe=""),
                address=credentials.address,
                port=credentials.port,
                database=credentials.database,
            )
        )
        self.mongo_database = self.mongo_client[credentials.database]

        self.reload()

    def shutdown(self) -> None:
        self.mongo_client.close()
        self.forget_sections()

    def discover_names(self, consumer: Callable[[str], None]) -> None:
        """Lists the database's collection names, skipping the MongoDB-internal ones."""
        for name in self.mongo_database.list_collection_names():
            if name in _FORBIDDEN:
                continue
            consumer(name)

    def construct_section(self, name: str, config: SectionConfig) -> AbstractCachedDatabaseSection:
        return MongoDBDatabaseSection(self.mongo_database, name, config)

    def drop_section_remote(self, name: str) -> None:
        self.mongo_database[name].drop()
